File: tests/Firewall.Native.RegressionTests.cpp
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.

// Standalone white-box ABI/codec regression runner; no persistent COM writes.
#include "WinTraceForge.Firewall.Native.cpp"
#include <iostream>

namespace
{
    int assertions = 0;
    void Assert(bool condition, const char* message)
    {
        ++assertions;
        if (!condition) { throw std::runtime_error(message); }
    }
    template<typename F> void Invalid(F action, const char* message)
    {
        Assert(FirewallNative::Guard(action) == E_INVALIDARG, message);
    }
    HRESULT Invoke(void* session, uint32_t operation, const std::vector<unsigned char>& packet)
    {
        void* output = nullptr;
        uint32_t length = 0;
        const HRESULT status = FirewallNativeInvoke(session, operation, packet.data(),
            static_cast<uint32_t>(packet.size()), &output, &length);
        if (FAILED(status)) { Assert(output == nullptr && length == 0, "failure outputs initialized"); }
        else { Assert(output != nullptr && length >= 4, "success outputs initialized"); }
        FirewallNativeFree(output);
        return status;
    }

    // Intercept the exact Rules.Add boundary; never forward to the real policy collection.
    struct AddProbe : INetFwRules
    {
        ULONG references = 1;
        unsigned calls = 0;
        bool applicationNull = false, serviceNull = false;
        std::wstring localPorts, remotePorts, application, service;
        long protocol = 0;
        HRESULT result = E_ACCESSDENIED;
        HRESULT STDMETHODCALLTYPE QueryInterface(REFIID id, void** value) override
        {
            if (!value) { return E_POINTER; }
            *value = nullptr;
            if (id != __uuidof(IUnknown) && id != __uuidof(IDispatch) && id != __uuidof(INetFwRules)) { return E_NOINTERFACE; }
            *value = static_cast<INetFwRules*>(this); AddRef(); return S_OK;
        }
        ULONG STDMETHODCALLTYPE AddRef() override { return ++references; }
        ULONG STDMETHODCALLTYPE Release() override
        {
            const ULONG left = --references;
            if (!left) { delete this; }
            return left;
        }
        HRESULT STDMETHODCALLTYPE GetTypeInfoCount(UINT*) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE GetTypeInfo(UINT, LCID, ITypeInfo**) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE GetIDsOfNames(REFIID, LPOLESTR*, UINT, LCID, DISPID*) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE Invoke(DISPID, REFIID, LCID, WORD, DISPPARAMS*, VARIANT*, EXCEPINFO*, UINT*) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE get_Count(long*) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE Remove(BSTR) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE Item(BSTR, INetFwRule**) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE get__NewEnum(IUnknown**) override { return E_NOTIMPL; }
        HRESULT STDMETHODCALLTYPE Add(INetFwRule* rule) override
        {
            using namespace FirewallNative;
            ++calls;
            return Guard([&] {
                Bstr app, svc, local, remote;
                Check(rule->get_ApplicationName(&app.value)); Check(rule->get_ServiceName(&svc.value));
                Check(rule->get_LocalPorts(&local.value)); Check(rule->get_RemotePorts(&remote.value));
                Check(rule->get_Protocol(&protocol));
                applicationNull = app.value == nullptr; serviceNull = svc.value == nullptr;
                application = app.value ? app.value : L""; service = svc.value ? svc.value : L"";
                localPorts = local.value ? local.value : L""; remotePorts = remote.value ? remote.value : L"";
                Check(result);
            });
        }
    };

    void AddBoundary()
    {
        using namespace FirewallNative;
        Check(CoInitializeEx(nullptr, COINIT_MULTITHREADED));
        try
        {
            for (const bool inbound : {false, true})
            {
                Policy policy;
                auto* probe = new AddProbe();
                policy.rules.Attach(probe);
                Writer packet;
                const wchar_t* fields[] = {
                    L"WinTraceForge.Detached.AddBoundary", L"WinTraceForge.Firewall.v1",
                    L"Detached test only", L"", L"", L"*", L"192.10.10.1",
                    inbound ? L"443" : L"*", inbound ? L"*" : L"1234", L"All",
                    L"", L"", L"", L"", L""
                };
                for (const auto* field : fields)
                {
                    Bstr text; text.value = SysAllocString(field); Check(text.value ? S_OK : E_OUTOFMEMORY);
                    packet.String(text.value);
                }
                packet.Int(0);
                for (const uint32_t value : {6u, inbound ? 1u : 2u, 0u, 1u, 0u, 0u, 1u, 0u}) { packet.Int(value); }
                Writer prepared;
                policy.Execute(5, packet.bytes, prepared);
                Writer empty, response;
                Assert(Guard([&] { policy.Execute(7, empty.bytes, response); }) == E_ACCESSDENIED,
                    "actual Add boundary preserves injected persistence HRESULT");
                Assert(probe->calls == 1, "operation 7 invokes collection exactly once");
                Assert(probe->applicationNull, "unspecified application remains NULL at Add boundary");
                Assert(probe->serviceNull, "unspecified service remains NULL at Add boundary");
                Assert(probe->protocol == 6, "protocol preserved at Add boundary");
                Assert(probe->localPorts == (inbound ? L"443" : L"*"), "local port scope at Add boundary");
                Assert(probe->remotePorts == (inbound ? L"*" : L"1234"), "remote port scope at Add boundary");
            }
        }
        catch (...) { CoUninitialize(); throw; }
        CoUninitialize();
    }
    void Codec()
    {
        using namespace FirewallNative;
        Writer output;
        output.String(nullptr);
        Bstr empty;
        empty.value = SysAllocString(L"");
        output.String(empty.value);
        Bstr unicode;
        unicode.value = SysAllocString(L"Ethernet-\x7f51\x7edc");
        output.String(unicode.value);
        Reader reader(output.bytes);
        Assert(reader.String().isNull, "null BSTR preserved");
        Text text = reader.String();
        Assert(!text.isNull && text.value.empty(), "empty BSTR preserved");
        Assert(reader.String().value == unicode.value, "UTF16 preserved");
        reader.End();
        Invalid([&] { reader.Int(); }, "truncated integer");
        Writer malformed;
        malformed.Int(MaxString + 1);
        Invalid([&] { Reader value(malformed.bytes); value.String(); }, "huge string rejected");
        malformed.bytes.resize(4);
        malformed.Int(2);
        Invalid([&] { Reader value(malformed.bytes); value.String(); }, "truncated string rejected");
        malformed.bytes.resize(4);
        malformed.Int(2);
        const wchar_t nul[] = {L'a', 0};
        malformed.Raw(nul, sizeof(nul));
        Invalid([&] { Reader value(malformed.bytes); value.String(); }, "embedded NUL rejected");
        malformed.bytes.resize(4);
        malformed.Int(2);
        Invalid([&] { Reader value(malformed.bytes); value.Boolean(); }, "noncanonical Boolean rejected");
        Invalid([&] { Reader value(malformed.bytes); value.End(); }, "trailing data rejected");
        Variant list;
        list.value.vt = VT_ARRAY | VT_VARIANT;
        SAFEARRAYBOUND bound = {2, 7};
        list.value.parray = SafeArrayCreate(VT_VARIANT, 1, &bound);
        Assert(list.value.parray != nullptr, "variant array allocation");
        Variant item;
        item.value.vt = VT_BSTR;
        item.value.bstrVal = SysAllocString(L"interface-7");
        LONG index = 7;
        Check(SafeArrayPutElement(list.value.parray, &index, &item.value));
        index = 8;
        Check(SafeArrayPutElement(list.value.parray, &index, &item.value));
        Writer strings;
        strings.Strings(list.value);
        Reader arrayReader(strings.bytes);
        Assert(arrayReader.Int() == 2, "nonzero lower bound count");
        Assert(arrayReader.String().value == L"interface-7", "variant array first");
        Assert(arrayReader.String().value == L"interface-7", "variant array second");
        arrayReader.End();
        Variant invalid;
        invalid.value.vt = VT_I4;
        Invalid([&] { Writer value; value.Strings(invalid.value); }, "non-array not defaulted");
        Variant wrongItem;
        wrongItem.value.vt = VT_I4;
        Check(SafeArrayPutElement(list.value.parray, &index, &wrongItem.value));
        Invalid([&] { Writer value; value.Strings(list.value); }, "non-string array item refused");
        Variant bstrList;
        bstrList.value.vt = VT_ARRAY | VT_BSTR;
        bstrList.value.parray = SafeArrayCreateVector(VT_BSTR, 0, 1);
        index = 0;
        Check(SafeArrayPutElement(bstrList.value.parray, &index, unicode.value));
        Writer bstrWriter;
        bstrWriter.Strings(bstrList.value);
        Reader bstrReader(bstrWriter.bytes);
        Assert(bstrReader.Int() == 1 && bstrReader.String().value == unicode.value, "BSTR array supported");
    }
    void Parameters()
    {
        using namespace FirewallNative;
        Assert(FirewallNativeOpen(nullptr) == E_POINTER, "null open output");
        Assert(FirewallNativeClose(nullptr) == E_HANDLE, "null session");
        Assert(FirewallNativeClose(reinterpret_cast<void*>(123456789)) == E_HANDLE, "unknown session");
        void* session = nullptr;
        Check(FirewallNativeOpen(&session));
        try
        {
            Writer packet;
            Assert(Invoke(session, 1, packet.bytes) == S_OK, "read current profile");
            Assert(Invoke(session, 2, packet.bytes) == S_OK, "read local modify state");
            Assert(Invoke(session, 3, packet.bytes) == S_OK, "read profiles");
            Assert(Invoke(session, 0, packet.bytes) == E_INVALIDARG, "invalid operation zero");
            Assert(Invoke(session, 9, packet.bytes) == E_INVALIDARG, "invalid operation upper");
            Assert(Invoke(reinterpret_cast<void*>(123456789), 1, packet.bytes) == E_HANDLE, "unknown invoke handle");
            Assert(Invoke(session, 7, packet.bytes) == HRESULT_FROM_WIN32(ERROR_INVALID_STATE),
                "Add without detached rule stops before Rules.Add");
            packet.Int(UINT32_MAX);
            Assert(Invoke(session, 4, packet.bytes) == E_INVALIDARG, "null lookup name");
            Assert(Invoke(session, 8, packet.bytes) == E_INVALIDARG, "null remove name rejected before Rules.Remove");
            packet.bytes.resize(4);
            packet.Int(0);
            Assert(Invoke(session, 4, packet.bytes) == E_INVALIDARG, "empty lookup name");
            packet.bytes.resize(4);
            packet.Int(15);
            Assert(Invoke(session, 4, packet.bytes) == E_INVALIDARG, "truncated name");
            Assert(Invoke(session, 5, packet.bytes) == E_INVALIDARG, "truncated rule");
            packet.bytes.resize(3);
            Assert(Invoke(session, 1, packet.bytes) == E_INVALIDARG, "short packet");
            packet.bytes.assign(4, 0);
            Assert(Invoke(session, 1, packet.bytes) == E_INVALIDARG, "bad magic");
            void* output = nullptr;
            uint32_t length = 0;
            Assert(FirewallNativeInvoke(session, 1, nullptr, 4, &output, &length) == E_INVALIDARG, "null input");
            Assert(FirewallNativeInvoke(session, 1, packet.bytes.data(), 4, nullptr, &length) == E_POINTER, "null output");
            Assert(FirewallNativeInvoke(session, 1, packet.bytes.data(), 4, &output, nullptr) == E_POINTER, "null length");
            Assert(FirewallNativeInvoke(session, 1, packet.bytes.data(), 0xffffffffu, &output, &length) == E_INVALIDARG, "oversized packet");
        }
        catch (...) { FirewallNativeClose(session); throw; }
        Check(FirewallNativeClose(session));
        Assert(FirewallNativeClose(session) == E_HANDLE, "double close");
        Writer packet;
        Assert(Invoke(session, 1, packet.bytes) == E_HANDLE, "closed handle invocation");
        FirewallNativeFree(nullptr);
    }
}
int main()
{
    try
    {
        Codec();
        Parameters();
        AddBoundary();
        std::cout << "PASS: " << assertions << " native assertions; no persistent COM writes.\n";
        return 0;
    }
    catch (const FirewallNative::Failure& error)
    { std::cerr << "HRESULT: " << std::hex << error.code << "\n"; return 1; }
    catch (const std::exception& error) { std::cerr << error.what() << "\n"; return 1; }
}

File: tests/Firewall.Native.RegressionTests.cs
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.

using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Threading;

// Compile alongside WinTraceForge*.cs with /main:FirewallNativeRegressionTests.
// The only write-like operations exercised are detached object setters. Never Add/Remove.
internal static class FirewallNativeRegressionTests
{
    private static int assertions;
    private static void Assert(bool value, string message)
    {
        ++assertions;
        if (!value) { throw new Exception(message); }
    }
    private static void Expect<T>(Action action, string message) where T : Exception
    {
        try { action(); }
        catch (T) { Assert(true, message); return; }
        throw new Exception("Expected " + typeof(T).Name + ": " + message);
    }
    private static FirewallRuleData Expected(int protocol)
    {
        FirewallOptions options = FirewallModule.Parse(new string[] { "rule", "add", "--id",
            Guid.NewGuid().ToString("D"), "--remote-address", protocol == 6 ? "192.0.2.15" : "2001:db8::1",
            "--remote-port", "443", "--protocol", protocol == 6 ? "tcp" : "udp", "--profiles", "private" });
        options.Direction = protocol == 6 ? 2 : 1;
        options.Action = protocol == 6 ? 0 : 1;
        options.LocalPort = protocol == 6 ? 0 : 12345;
        options.Program = protocol == 6 ? "" : @"C:\Windows\System32\notepad.exe";
        return FirewallModule.ExpectedRule(options, 2);
    }
    private static void CompareFields(object expected, object actual)
    {
        foreach (FieldInfo field in expected.GetType().GetFields(BindingFlags.Instance | BindingFlags.NonPublic))
        {
            object left = field.GetValue(expected), right = field.GetValue(actual);
            string[] strings = left as string[];
            if (strings != null)
            {
                string[] other = right as string[];
                Assert(other != null && strings.Length == other.Length, field.Name + " count");
                for (int i = 0; i < strings.Length; ++i) { Assert(strings[i] == other[i], field.Name + " element"); }
            }
            else { Assert(object.Equals(left, right), "Full field comparison: " + field.Name); }
        }
    }
    private static List<string> ExistingNames()
    {
        object policy = null, rules = null, enumeration = null;
        object[] item = new object[1];
        List<string> names = new List<string>();
        try
        {
            policy = Activator.CreateInstance(Type.GetTypeFromProgID("HNetCfg.FwPolicy2", true));
            rules = Get(policy, "Rules");
            IntPtr pointer;
            int status = ((ComFirewallBackend.IFirewallRulesEnumerator)rules).GetNewEnum(out pointer);
            try
            {
                Marshal.ThrowExceptionForHR(status);
                enumeration = Marshal.GetTypedObjectForIUnknown(pointer, typeof(ComFirewallBackend.IFirewallEnumVariant));
            }
            finally { if (pointer != IntPtr.Zero) { Marshal.Release(pointer); } }
            while (names.Count < 5)
            {
                status = ((ComFirewallBackend.IFirewallEnumVariant)enumeration).Next(1, item, IntPtr.Zero);
                try
                {
                    if (status == 1) { break; }
                    Marshal.ThrowExceptionForHR(status);
                    Assert(status == 0, "enumerator status");
                    int protocol = (int)Get(item[0], "Protocol");
                    if (protocol == 6 || protocol == 17)
                    {
                        string name = (string)Get(item[0], "Name");
                        if (!names.Contains(name)) { names.Add(name); }
                    }
                }
                finally { Release(item[0]); item[0] = null; }
            }
            return names;
        }
        finally { Release(item[0]); Release(enumeration); Release(rules); Release(policy); }
    }
    private static object Get(object target, string property)
    {
        return target.GetType().InvokeMember(property, BindingFlags.GetProperty, null, target, null);
    }
    private static void Release(object target)
    {
        if (target != null && Marshal.IsComObject(target)) { Marshal.FinalReleaseComObject(target); }
    }
    private static void ReadOnly()
    {
        using (IFirewallBackend com = new ComFirewallBackend())
        using (IFirewallBackend native = new NativeFirewallBackend())
        {
            Assert(com.CurrentProfiles == native.CurrentProfiles, "current profiles");
            Assert(com.LocalPolicyModifyState == native.LocalPolicyModifyState, "modify state");
            IList<FirewallProfileData> left = com.ReadProfiles(), right = native.ReadProfiles();
            Assert(left.Count == 3 && right.Count == 3, "all profiles");
            for (int i = 0; i < 3; ++i) { CompareFields(left[i], right[i]); }
            string missing = "WinTraceForge.Native.Absent." + Guid.NewGuid().ToString("D");
            Assert(com.FindByName(missing).Count == 0 && native.FindByName(missing).Count == 0, "absent lookup");
            List<string> names = ExistingNames();
            Assert(names.Count != 0, "at least one existing TCP/UDP rule available");
            foreach (string name in names)
            {
                IList<FirewallRuleData> comRules = com.FindByName(name), nativeRules = native.FindByName(name);
                Assert(comRules.Count > 0 && comRules.Count == nativeRules.Count, "existing rule count");
                for (int i = 0; i < comRules.Count; ++i) { CompareFields(comRules[i], nativeRules[i]); }
                Assert(native.FindByName(name.ToUpperInvariant()).Count == nativeRules.Count, "case insensitive collision lookup");
            }
            Console.WriteLine("Compared all DTO fields for {0} existing names and three profiles.", names.Count);
        }
    }
    private static void Detached()
    {
        FirewallOptions inbound = FirewallModule.Parse(new string[] { "rule", "add", "--id",
            Guid.NewGuid().ToString("D"), "--direction", "in", "--remote-address", "192.0.2.15",
            "--local-port", "443", "--profiles", "private" });
        FirewallRuleData inboundRule = FirewallModule.ExpectedRule(inbound, 2);
        Assert(inboundRule.Direction == 1 && inboundRule.LocalPorts == "443" && inboundRule.RemotePorts == "*",
            "inbound local443 with omitted source port maps to wildcard");
        using (IFirewallBackend com = new ComFirewallBackend())
        using (IFirewallBackend native = new NativeFirewallBackend())
        {
            Assert(com.FindByName(inboundRule.Name).Count == 0 && native.FindByName(inboundRule.Name).Count == 0,
                "inbound wildcard rule absent before preparation");
            com.PrepareAdd(inboundRule);
            native.PrepareAdd(inboundRule);
            Assert(com.FindByName(inboundRule.Name).Count == 0 && native.FindByName(inboundRule.Name).Count == 0,
                "inbound local443 remote wildcard detached readback succeeds without persistence");
        }
        foreach (int protocol in new int[] { 6, 17 })
        {
            FirewallRuleData rule = Expected(protocol);
            using (IFirewallBackend com = new ComFirewallBackend())
            using (IFirewallBackend native = new NativeFirewallBackend())
            {
                Assert(native.FindByName(rule.Name).Count == 0, "detached absent before");
                com.PrepareAdd(rule);
                native.PrepareAdd(rule);
                Assert(com.FindByName(rule.Name).Count == 0 && native.FindByName(rule.Name).Count == 0,
                    "detached setters never persisted");
                Expect<FirewallRefusalException>(delegate { native.PrepareAdd(rule); }, "duplicate preparation");
            }
        }
        using (IFirewallBackend native = new NativeFirewallBackend())
        {
            Expect<FirewallRefusalException>(delegate { native.Add(); }, "Add without prepare refused before COM call");
            FirewallRuleData invalid = Expected(6);
            invalid.Protocol = -1;
            Expect<COMException>(delegate { native.PrepareAdd(invalid); }, "invalid protocol HRESULT");
            invalid = Expected(6);
            invalid.RemoteAddresses = "not-an-ip-address";
            Expect<COMException>(delegate { native.PrepareAdd(invalid); }, "invalid address HRESULT");
            foreach (string property in new string[] { "LocalAppPackageId", "LocalUserOwner", "LocalUserAuthorizedList",
                "RemoteUserAuthorizedList", "RemoteMachineAuthorizedList" })
            {
                invalid = Expected(6);
                typeof(FirewallRuleData).GetField(property, BindingFlags.NonPublic | BindingFlags.Instance).SetValue(invalid, "must-not-be-lost");
                Expect<FirewallRefusalException>(delegate { native.PrepareAdd(invalid); }, property + " survives preflight");
            }
            invalid = Expected(6);
            invalid.Interfaces = new string[] { "restricted-interface" };
            Expect<FirewallRefusalException>(delegate { native.PrepareAdd(invalid); }, "interface restriction refusal");
            invalid = Expected(6);
            invalid.SecureFlags = 1;
            Expect<FirewallRefusalException>(delegate { native.PrepareAdd(invalid); }, "IPsec restriction refusal");
            invalid = Expected(6);
            invalid.EdgeTraversalOptions = 1;
            Expect<FirewallRefusalException>(delegate { native.PrepareAdd(invalid); }, "edge restriction refusal");
            invalid = Expected(6);
            invalid.EdgeTraversal = true;
            Expect<FirewallRefusalException>(delegate { native.PrepareAdd(invalid); }, "edge traversal refusal");
            native.PrepareAdd(Expected(6));
            Assert(true, "failed preparations leave session reusable");
        }
    }
    private static void Codec()
    {
        FirewallRuleData original = Expected(6);
        original.Name = "UTF16-\u00e9-\ud83d\ude80-\ud800";
        original.Interfaces = new string[] { "Ethernet", "\u7f51\u7edc", "", null };
        original.LocalAppPackageId = "package-SID";
        original.LocalUserOwner = null;
        original.LocalUserAuthorizedList = "D:(A;;CC;;;WD)";
        original.RemoteUserAuthorizedList = "remote-user";
        original.RemoteMachineAuthorizedList = "remote-machine";
        original.EdgeTraversalOptions = 2;
        original.SecureFlags = unchecked((int)0x80000001);
        original.EdgeTraversal = true;
        Type backend = typeof(NativeFirewallBackend);
        MethodInfo encode = backend.GetMethod("RulePacket", BindingFlags.Static | BindingFlags.NonPublic);
        byte[] packet = (byte[])encode.Invoke(null, new object[] { original });
        Type readerType = backend.GetNestedType("PacketReader", BindingFlags.NonPublic);
        using (IDisposable reader = (IDisposable)Activator.CreateInstance(readerType,
            BindingFlags.Instance | BindingFlags.NonPublic, null, new object[] { packet }, null))
        {
            FirewallRuleData copy = (FirewallRuleData)readerType.GetMethod("Rule", BindingFlags.Instance | BindingFlags.NonPublic).Invoke(reader, null);
            CompareFields(original, copy);
            readerType.GetMethod("End", BindingFlags.Instance | BindingFlags.NonPublic).Invoke(reader, null);
        }

        foreach (byte[] bad in new byte[][] { new byte[0], new byte[3], new byte[4] })
        {
            try
            {
                Activator.CreateInstance(readerType, BindingFlags.Instance | BindingFlags.NonPublic, null, new object[] { bad }, null);
                throw new Exception("Malformed packet was accepted.");
            }
            catch (TargetInvocationException error) { Assert(error.InnerException is FirewallRefusalException, "malformed packet rejected"); }
        }
    }

    private static void OptionalDefaults()
    {
        foreach (int protocol in new[] { 6, 17 })
        {
            FirewallRuleData expected = Expected(protocol);
            using (ComFirewallBackend com = new ComFirewallBackend())
            using (NativeFirewallBackend native = new NativeFirewallBackend())
            {
                com.PrepareAdd(expected);
                object comRule = typeof(ComFirewallBackend).GetField("preparedRule",
                    BindingFlags.Instance | BindingFlags.NonPublic).GetValue(com);
                Assert(Get(comRule, "ServiceName") == null, "COM unspecified service stays NULL, not empty");
                Assert(protocol == 6 ? Get(comRule, "ApplicationName") == null :
                    (string)Get(comRule, "ApplicationName") == expected.ApplicationName,
                    "COM omitted/explicit application preserved");
                Type backend = typeof(NativeFirewallBackend);
                byte[] packet = (byte[])backend.GetMethod("RulePacket", BindingFlags.Static | BindingFlags.NonPublic)
                    .Invoke(null, new object[] { expected });
                using (IDisposable reader = (IDisposable)backend.GetMethod("Invoke", BindingFlags.Instance | BindingFlags.NonPublic)
                    .Invoke(native, new object[] { (uint)5, packet }))
                {
                    FirewallRuleData actual = (FirewallRuleData)reader.GetType()
                        .GetMethod("Rule", BindingFlags.Instance | BindingFlags.NonPublic).Invoke(reader, null);
                    reader.GetType().GetMethod("End", BindingFlags.Instance | BindingFlags.NonPublic).Invoke(reader, null);
                    Assert(actual.ServiceName == null, "native unspecified service stays NULL, not empty");
                    Assert(protocol == 6 ? actual.ApplicationName == null : actual.ApplicationName == expected.ApplicationName,
                        "native omitted/explicit application preserved");
                    Assert(FirewallModule.Mismatches(expected, actual).Count == 0, "optional default change preserves all requested scope");
                }
                Assert(com.FindByName(expected.Name).Count == 0 && native.FindByName(expected.Name).Count == 0,
                    "optional-default checks never persist a rule");
            }
        }
        MethodInfo operation = typeof(NativeFirewallBackend).GetMethod("OperationName", BindingFlags.Static | BindingFlags.NonPublic);
        Assert((string)operation.Invoke(null, new object[] { (uint)7 }) == "INetFwRules::Add", "Add failure names actual API");
        Assert((string)operation.Invoke(null, new object[] { (uint)8 }) == "INetFwRules::Remove", "Remove failure names actual API");
        try
        {
            typeof(NativeFirewallBackend).GetMethod("Throw", BindingFlags.Static | BindingFlags.NonPublic)
                .Invoke(null, new object[] { unchecked((int)0x80070057), "INetFwRules::Add" });
            throw new Exception("Native error was swallowed.");
        }
        catch (TargetInvocationException error)
        {
            COMException failure = error.InnerException as COMException;
            Assert(failure != null && failure.HResult == unchecked((int)0x80070057),
                "named Add error preserves original HRESULT");
            Assert(failure.Message.Contains("INetFwRules::Add") && !failure.Message.Contains("0x80070057"),
                "HRESULT left for shared reporter, not duplicated in message");
        }
    }
    private static void Lifetime()
    {
        NativeFirewallBackend native = new NativeFirewallBackend();
        Exception failure = null;
        Thread thread = new Thread(delegate()
        {
            try { Assert(native.ReadProfiles().Count == 3, "cross-thread read"); }
            catch (Exception error) { failure = error; }
        });
        thread.Start();
        thread.Join();
        if (failure != null) { throw failure; }
        Expect<ArgumentException>(delegate { native.FindByName(""); }, "empty name");
        Expect<ArgumentException>(delegate { native.FindByName("prefix\0suffix"); }, "embedded NUL name");
        Expect<ArgumentException>(delegate { native.FindByName(new string('a', 1024 * 1024 + 1)); }, "oversized name");
        Expect<ArgumentNullException>(delegate { native.PrepareAdd(null); }, "null rule");
        native.Dispose();
        native.Dispose();
        Expect<ObjectDisposedException>(delegate { native.ReadProfiles(); }, "disposed access");
        Expect<ObjectDisposedException>(delegate { native.Add(); }, "disposed Add");
        for (int i = 0; i < 4; ++i) { using (NativeFirewallBackend value = new NativeFirewallBackend()) { Assert(value.CurrentProfiles > 0, "reopen"); } }
    }
    private static int Main()
    {
        try
        {
            Codec();
            ReadOnly();
            Detached();
            OptionalDefaults();
            Lifetime();
            Console.WriteLine("PASS: {0} assertions; no Rules.Add/Remove or policy setter invoked.", assertions);
            return 0;
        }
        catch (Exception error) { Console.Error.WriteLine(error); return 1; }
    }
}
